src/bin/sccache-dist/main.rs

```diff
@@ -318,6 +321,123 @@ fn parse() -> Result<Command> {
+    let pkcs1 = kid_to_pkcs1.get(&kid).chain_err(|| "kid not found in jwks")?;
+    let mut validation = jwt::Validation::new(header.alg);
+    validation.set_audience(&audience);
+    validation.iss = Some(issuer.to_owned());
+    #[derive(Deserialize)]
+    struct Claims {}
+    // Decode the JWT, discarding any claims - we just care about validity
+    let _tokendata = jwt::decode::<Claims>(token, pkcs1, &validation)
+        .chain_err(|| "Unable to validate and decode jwt")?;
+    Ok(())
+}
+
+// https://infosec.mozilla.org/guidelines/iam/openid_connect#session-handling
+const MOZ_SESSION_TIMEOUT: Duration = Duration::from_secs(60 * 15);
+
+// Mozilla-specific check by forwarding the token onto person api
+fn check_mozilla(auth_cache: &Mutex<HashMap<String, Instant>>, client: &reqwest::Client, token: &str) -> Result<()> {
+    #[derive(Deserialize)]
+    struct MozillaToken {
+        exp: u64,
+        sub: String,
+    }
+    let unsafe_token = jwt::dangerous_unsafe_decode::<MozillaToken>(token).chain_err(|| "Unable to decode jwt")?;
+    let user = unsafe_token.claims.sub;
+    if UNIX_EPOCH + Duration::from_secs(unsafe_token.claims.exp) < SystemTime::now() {
+        bail!("JWT expired")
+    }
+    // If the token is cached and not expired, return it
+    {
+        let mut auth_cache = auth_cache.lock().unwrap();
+        if let Some(cached_at) = auth_cache.get(token) {
+            if cached_at.elapsed() < MOZ_SESSION_TIMEOUT {
+                return Ok(())
+            }
+        }
+        auth_cache.remove(token);
+    }
+    // Ask person api about groups (as a side effect, checking the JWT)
+    // https://github.com/mozilla-iam/cis/blob/master/docs/rfcs/UserProfilesv2.md
+    let url = reqwest::Url::parse("https://person-api.sso.mozilla.com/v2/profile/").unwrap().join(&user)
+        .chain_err(|| format!("Could not create person api url for {}", user))?;
+    let header = reqwest::header::Authorization(reqwest::header::Bearer { token: token.to_owned() });
+    let mut res = client.get(url.clone()).header(header).send().unwrap();
+    if !res.status().is_success() {
+        bail!("JWT forwarded to {} returned {} {}", url, res.status().as_u16(), res.status());
+    }
+    #[derive(Deserialize)]
+    struct CISProfileV2AccessInformationLDAPItem {
+        name: String,
+    }
+    #[derive(Deserialize)]
+    struct CISProfileV2AccessInformationLDAP {
+        value: Vec<CISProfileV2AccessInformationLDAPItem>,
+    }
+    #[derive(Deserialize)]
+    struct CISProfileV2AccessInformation {
+        ldap: CISProfileV2AccessInformationLDAP,
+    }
+    #[derive(Deserialize)]
+    struct CISProfileV2 {
+        access_information: CISProfileV2AccessInformation,
+    }
+    let profile: CISProfileV2 = res.json().chain_err(|| "Cannot decode CIS V2 profile from person API")?;
+    let is_allowed = profile.access_information.ldap.value.into_iter().any(|item| item.name == "hris_dept_firefox");
```

Comment:
I'm not 100% confident that this is the check we want here, but maybe it's close enough for now? I don't know if "is a MoCo employee" would be considered too broad? We should probably ask the security folks for clarification.